basetest.pm

```diff
@@ -645,19 +645,22 @@ sub parse_serial_output_qemu {
         chomp $line;
         # only two keys allowed
         for my $type (qw(soft hard)) {
-            for my $regexp (@{$failures->{$type}}) {
-                # If you want to match a simple string please be sure that you create it with quotemeta
-                if ($line =~ /$regexp/) {
-                    my $fail_type = 'softfail';
-                    if ($type eq 'hard') {
-                        $die       = 1;
-                        $fail_type = 'fail';
+            for my $regexp_table ($failures->{$type}) {
+                for my $msg (keys %$regexp_table) {
+                    for my $regexp (@{$regexp_table->{$msg}}) {
+                        # If you want to match a simple string please be sure that you create it with quotemeta
+                        if ($line =~ /$regexp/) {
```

Comment:
I don't really get how we ended up with O(n^3) here (considering first loop is a constant). What's about changing structure, I also think it will be enough to have one regexp per message, as for OR scenarios regexp can be easily written. Reason to have array in previous implementation is to be able to match multiple pattern for different use-cases.
So I suggest following structure:
```perl
my %failures = (
    soft => [
            { message => '1', pattern => qr/Pattern1/  },
            { message => '2', pattern => qr/Pattern2/  },
    ],
    hard => [
            { message => '3', pattern => [qr/Pattern3/] },
    ],
);
```
Then we can iterate like:
```perl
for my $type (qw(soft hard)) {
    for my $regexp_table (@{$failures{$type}}) {
        print $regexp_table->{message} . " " . $regexp_table->{pattern} . "\n";
    }
}
```
Or using same approach with message as a key and pattern as value and having something like:
```perl
while(my($message, $pattern) = each %hash) { ... }
```

WDYT?

Reply:
@rwx788 That's a good idea. But, imho, if we want to make the code clearer i think we should use the regular expression as key. Then we could loop the file with something like:

`
while(<>)  {
 my $match = $hash{$_};
 print "found $match->{$message} with $match->{type}\n" if $match;
}
`

If i'm correct this supports all the possible cases without the need to change anything in the future, if we want to extend.

Although this would bring a new cpan dep (Tie::RegexpHash). :-(

Reply:
Yeah, main thing which we got on top is array or regexps per message, but it will be enough to have 1 to 1, whether message or regexp is the key. Having same regexp twice with different messages and behavior is indeed wrong. 4 nested for loops looks like overkill. On top we can set type of failure as a property of failure, having just array of hashes in the end. It only depends if we plan to extend it even further. But I like your proposal, for now it's perfect (except extra dependency, but that's minor).
E.g.
```perl
my %failures = (
    [
            { type => soft, message => '1', pattern => qr/Pattern1/ },
            { type => soft, message => '2', pattern => qr/Pattern2/ },
            { type => hard, message => '3', pattern => qr/Pattern3/ },
    ]
);
```

Reply:
I made it work with the first approach mentioned by @rwx788 changing the internal structure and removing some loops. Besides I added a hash for not repeating matches when recording the results as suggested by @dasantiago.
Verification run: http://dhcp254.suse.cz/tests/1723#